cmd/virt-launcher/virt-launcher.go

```diff
@@ -475,9 +475,12 @@ func main() {
 			finalShutdownCallback,
 			gracefulShutdownCallback)
 
+		killChan := make(chan struct{})
+		go watchForVMIKillSignal(killChan)
+
 		// This is a wait loop that monitors the qemu pid. When the pid
 		// exits, the wait loop breaks.
-		mon.RunForever(*qemuTimeout, signalStopChan)
+		mon.RunForever(*qemuTimeout, signalStopChan, killChan)
```

Comment:
@davidvossel you had originally wrote this code so I am interested in your opinion since maybe this option isn't desired and I should fix this a different way. 

The reason for this change is the following scenario was happening regularly: 

- qemu came up 
- kill signal was recieved 
- monitor had not yet picked up on the process coming up

The pod would then get stuck in waiting in the graceful shutdown callback for this. The following log then repeats until the graceful shutdown period expires. Even though the process had already been up, it was just killed immediately. 

```
{"component":"virt-launcher","level":"info","msg":"Still missing PID for 45bc6ca0-0c90-5b4c-8186-d47a5217c94e, Process 45bc6ca0-0c90-5b4c-8186-d47a5217c94e not found in /proc","pos":"monitor.go:122","timestamp":"2021-06-04T00:53:32.696352Z"}
{"component":"virt-launcher","level":"info","msg":"Still missing PID for 45bc6ca0-0c90-5b4c-8186-d47a5217c94e, Process 45bc6ca0-0c90-5b4c-8186-d47a5217c94e not found in /proc","pos":"monitor.go:122","timestamp":"2021-06-04T00:53:33.698114Z"}
{"component":"virt-launcher","level":"info","msg":"Still missing PID for 45bc6ca0-0c90-5b4c-8186-d47a5217c94e, Process 45bc6ca0-0c90-5b4c-8186-d47a5217c94e not found in /proc","pos":"monitor.go:122","timestamp":"2021-06-04T00:53:34.695479Z"}
```

Reply:
so this killChan is only needed to close the gap the Domain initially being detected as existing in libvirt and the domain's process being detected by the monitor function?

If so, I'm trying to think if there's another way to achieve the same thing.

For instance, what if the KillVirtualMachine function raised SigTerm on itself (maybe that's possible, not 100% sure)? That would get reported back to the monitor loop via the `signalStopChan` which we could probably use to early terminate the loop.

Reply:
> so this killChan is only needed to close the gap the Domain initially being detected as existing in libvirt and the domain's process being detected by the monitor function?

correct

> For instance, what if the KillVirtualMachine function raised SigTerm on itself (maybe that's possible, not 100% sure)? That would get reported back to the monitor loop via the signalStopChan which we could probably use to early terminate the loop.

This already happens. The `signalStopChan` is closed and then initiates the graceful shut down [1]. The problem is that we never detected the domain and then we killed it so graceful shutdown keeps waiting for it for the length of the graceful shutdown period. I'm not sure how we tell the difference between the channel closing for this reason and the other reasons [2] that we do want the graceful shutdown to occur. Unless we send the signal/reason to the `signalStopChan` and SigTerm always skips graceful shutdown? 

[1] https://github.com/kubevirt/kubevirt/pull/5723/files#diff-335d247ac098e45fa4a39f21e175ac0cc2313278dfee4a04e9a5f9fe6d7867c4R178-R184 
[2] https://github.com/kubevirt/kubevirt/pull/5723/files#diff-335d247ac098e45fa4a39f21e175ac0cc2313278dfee4a04e9a5f9fe6d7867c4R178-R184

Reply:
> This already happens. The signalStopChan is closed and then initiates the graceful shut down [1]

ah okay, so what about something like this during pid detection. 

```
diff --git a/pkg/virt-launcher/monitor.go b/pkg/virt-launcher/monitor.go
index 42ef35e70..b58e9de38 100644
--- a/pkg/virt-launcher/monitor.go
+++ b/pkg/virt-launcher/monitor.go
@@ -128,6 +128,9 @@ func (mon *monitor) refresh() {
                        } else if expired {
                                log.Log.Infof("%s not found after grace period expired", mon.cmdlineMatchStr)
                                mon.isDone = true
+                       } else if mon.gracePeriodStartTime != 0 {
+                               log.Log.Infof("%s not found after shutdown initiated", mon.cmdlineMatchStr)
+                               mon.isDone = true
                        }
                        return
                }

```

Reply:
That's a much better solution :face_palm: I have updated with this change. 

I'm also wondering about what the `"should attempt graceful shutdown of Domain if no cluster wide equivalent exists"` test is testing. Is it suppose to be testing a Domain without an associated VMI? I can't get this test working with the change to the `isACPIEnabled` function but I also don't understand what it's testing.

Reply:
> Is it suppose to be testing a Domain without an associated VMI?

yes.  I believe this was before we had a finalizer on the VMI to ensure that guests would still be terminated even if the VMI object disappeared.

Today, under normal operation the VMI can't disappear unless the pod disappears due to a finalizer managed by virt-controller. It's technically possible to delete the VMI and orphan the pod ownership though, so maybe this logic needs to remain in virt-handler. 

> I can't get this test working with the change to the isACPIEnabled function but I also don't understand what it's testing.

oh no, so the `isACPIEnabled` function has been wrong this whole time! i think there's an easy change, i'll comment on that specific line.